src/simulator/halo_simulator.py

```diff
@@ -267,6 +267,43 @@ def simulated_venn_diagram_reach_by_spend(
         """
         raise NotImplementedError()
 
+    def _sample_venn_diagram(
+        self,
+        primitive_regions: Dict[int, List],
+        sample_size: int,
+        random_generator: np.random.Generator = np.random.default_rng(),
+    ) -> Dict[int, List]:
+        """Return primitive regions with sampled reaches.
+
+        Args:
+            primitive_regions:  Contains k+ reaches in the regions. The k+
+              reaches for a given region is given as a list r[] where r[k] is
+              the number of people who were reached AT LEAST k+1 times.
+            sample_size:  The total number of sampled reach from the primitive
+              regions.
+            random_generator:  An instance of numpy.random.Generator that is
+              used for generating samples from a multivariate hypergeometric
+              distribution.
+        Returns:
+            A dictionary in which each key are the binary representations of
+              each primitive region of the Venn diagram, and each value is a
+              list of the reach in the corresponding region.
+        """
+
+        region_repr_and_reach_pairs = [
+            (region_repr, kplus_reaches[0])
+            for region_repr, kplus_reaches in primitive_regions.items()
+        ]
+        region_repr_seq, reach_population = list(zip(*region_repr_and_reach_pairs))
+
+        sampled_reach = random_generator.multivariate_hypergeometric(
+            reach_population, sample_size
+        )
+
+        return {
+            region_repr: [r] for region_repr, r in zip(region_repr_seq, sampled_reach)
```

Comment:
I am thinking to make the value in the dictionary just reach instead of 1+ reaches (ie. [reach]) when all the Venn diagram changes are gathered.

Reply:
+1

Reply:
+1

Reply:
Thanks for the feedback, @jiayu-google and @matthewclegg . 

Since other functions like `_form_venn_diagram_regions`, `_generate_reach_points_from_venn_diagram`, and etc. also assumes 
```python 
primitive_regions: Dict[int, List]
```
Should I keep this one as what it is now, and then make the changes in another PR for all related functions?

Reply:
Good question. I don't think we need to change other places. 

The k+ reach of a primitive region or a subset of pubs is meaningful --- later we do want k+ reach to generate training points for frequency model. 

As for the `_sample_venn_diagram` method here, ideally we also want to include k+ reach; but, it is **infeasible**. (Precisely, the complete `primitive_regions: Dict[int, List]` does not contain enough info to obtain the sampled `Dict[int, List]`. To obtain the latter, we need info all the down from the raw DataSet.)

Reply:
I agree that k+ reaches will be useful in the future when we are ready. However, for now if we want to change the output type of `_sample_venn_diagram` to `Dict[int, int]`, the downstream functions need to change accordingly. For example, `_generate_reach_points_from_venn_diagram` is currently accepting `primitive_regions: Dict[int, **List**]`. If its input, which is the output from `_sample_venn_diagram`, changes to `Dict[int, **int**]`, we need to modify the code in `_generate_reach_points_from_venn_diagram` to accommodate the change.

My thought is:
`_form_venn_diagram_regions` still outputs regions with k+ reaches, so the functionality of obtaining k+ reaches still remains. Then let `_sampled_venn_diagram` output regions with sampled **reach**, and all the downstream functions use the regions with reach.

WDYT?

Reply:
Sounds like a good plan!

Reply:
Offline discussion with @matthewclegg: Matthew suggested to keep the type of `primitive_regions` as `Dict[int, List]` for now. After all the functionalities of Venn diagram query are done, we can come back to change the code.

Reply:
The type of `primitive_regions` will become `Dict[int, int]` during the refactoring.